test/integration/short/pool-simulator-tests.js

```diff
@@ -86,6 +91,98 @@ describe('pool', function () {
+        }
+        return retryPolicy.retryResult(undefined, true);
+      };
+
+      const client = new Client({
+        contactPoints: cluster.getContactPoints(),
+        policies: {
+          loadBalancing: new helper.OrderedLoadBalancingPolicy(),
+          retry: retryPolicy
+        },
+        pooling: {
+          coreConnectionsPerHost: {
+            [types.distance.local]: 2
+          },
+          maxRequestsPerConnection: 50
+        },
+        socketOptions: { defunctReadTimeoutThreshold: Number.MAX_SAFE_INTEGER }
+      });
+
+      return client.connect()
+        .then(() => firstHost = client.hosts.values()[0].address)
+        .then(() => client.execute('SELECT * FROM paused_on_first', null, { readTimeout: 200, isIdempotent: true }))
+        .then(() => assert.strictEqual(retryCount, 2))
+        .then(() => client.shutdown());
+    });
+
+    it('should retry on the next host when useCurrentHost is true and connections are not available', () => {
+      let firstHost;
+      const retryPolicy = policies.defaultRetryPolicy();
+      let retryPolicyCalled = 0;
+
+      retryPolicy.onRequestError = (info, c, err) => {
+        helper.assertInstanceOf(err, errors.OperationTimedOutError);
+        assert.notEqual(firstHost, undefined);
+        assert.strictEqual(err.host, firstHost);
+        retryPolicyCalled++;
+
+        // Use a safety mechanism to avoid retrying forever, in case the assertions fail
+        const useCurrentHost = info.nbRetry < 5;
+        return retryPolicy.retryResult(undefined, useCurrentHost);
+      };
+
+      const client = new Client({
+        contactPoints: cluster.getContactPoints(),
+        policies: {
+          loadBalancing: new helper.OrderedLoadBalancingPolicy(),
+          retry: retryPolicy
+        },
+        pooling: {
+          coreConnectionsPerHost: {
+            [types.distance.local]: 2
+          },
+          maxRequestsPerConnection: 50
+        },
+        socketOptions: { defunctReadTimeoutThreshold: Number.MAX_SAFE_INTEGER }
+      });
+
+      const promises = [];
+
+      return client.connect()
+        .then(() => {
+          firstHost = client.hosts.values()[0].address;
+
+          // Create 99 in-flight requests
+          promises.push.apply(promises, new Array(99).fill(0).map(() =>
+            client.execute('SELECT * FROM paused_on_first', null, { readTimeout: 2000 })));
+
+          // The pool will be busy after this request, so the retry will occur on the next host
+          const p2 = client.execute('SELECT * FROM paused_on_first', null, { readTimeout: 50, isIdempotent: true });
+          promises.push(p2);
+          return p2;
+        })
+        // The query will be retried on the next host
+        .then(rs => assert.strictEqual(rs.info.queriedHost, client.hosts.values()[1].address))
+        .then(() => Promise.all(promises))
+        .then(() => assert.strictEqual(retryPolicyCalled, 1))
```

Comment:
It would be good to have an server side verification. e.g. Checking the logs at simulacron to verify if the request was made a different (next) host. Wdyt?

Reply:
I think that the code is able to validate the queried host is probably good enough for the intent of the test.